File: scheme/ocaml/library.ml
open Ast

exception RuntimeError of string

let lambda : library_macro = fun args env -> match args with
    | List params::tl -> Closure(params, env, StatementList tl)
    | _ -> raise (RuntimeError "lambda: invalid arguments")

let define : library_macro = fun args env -> match args with
    | Var name::tl ->
        if Environment.mem env name then raise (RuntimeError "define: cannot redefine variable");
        let value = Eval.eval (StatementList tl) env in
        Environment.set env name value;
        value
    | _ -> raise (RuntimeError "define: invalid arguments")

let set : library_macro = fun args env -> match args with
    | Var name::tl ->
        let value = Eval.eval (StatementList tl) env in
        Environment.set env name value;
        value
    | _ -> raise (RuntimeError "set!: invalid arguments")

let letm : library_macro = fun args env -> match args with
    | List definitions::expr ->
        let new_env = Environment.new_with_parent env in
        let iterf definition = match definition with
            | List(Var name::tl) ->
                let value = Eval.eval (StatementList tl) env in
                Environment.set new_env name value
            | _ -> raise (RuntimeError "let: invalid arguments")
        in
        List.iter iterf definitions;
        Eval.eval (StatementList expr) new_env
    | _ -> raise (RuntimeError "let: invalid arguments")

let ifm : library_macro = fun args env -> match args with
    | [condition_expr; iftrue; iffalse] ->
        (match Eval.eval condition_expr env with
            | Bool false -> Eval.eval iffalse env
            | _ -> Eval.eval iftrue env)
    | _ -> raise (RuntimeError "if: invalid arguments")

let eval : library_macro = fun args env -> match args with
    | [expr] -> Eval.eval (Eval.eval expr env) env
    | _ -> raise (RuntimeError "eval: invalid arguments")

let defmacro : library_macro = fun args define_env -> match args with
    | Var name::List params::tl ->
        let new_macro : library_macro = fun args call_env ->
            let new_env = Environment.new_with_parent_and_meval define_env call_env in
            Eval.set_args params args new_env;
            Eval.eval (StatementList tl) new_env
        in
        Environment.set define_env name (LibraryMacro new_macro);
        LibraryMacro new_macro
    | _ -> raise (RuntimeError "defmacro: invalid arguments")

let meval : library_macro = fun args env -> match args, Environment.get_meval env with
    | [expr], Some env' -> Eval.eval (Eval.eval expr env) env'
    | _ -> raise (RuntimeError "meval: invalid arguments")

let print : library_function = fun args ->
    List.iter (fun arg -> Printf.printf "%s" (Ast.string_of_expr arg ^ "\n")) args;
    List []

let car : library_function = fun args -> match args with
    | [List(hd::_)] -> hd
    | _ -> raise (RuntimeError "car: invalid arguments")

let cdr : library_function = fun args -> match args with
    | [List(_::tl)] -> List tl
    | _ -> raise (RuntimeError "cdr: invalid arguments")

let cons : library_function = fun args -> match args with
    | [hd; List(tl)] -> List(hd::tl)
    | _ -> raise (RuntimeError "cons: invalid arguments")

let plus : library_function = fun args ->
    let numbers = List.map (fun arg -> match arg with
        | Integer n -> n
        | _ -> raise (RuntimeError "+: invalid arugments")) args in
    Integer (List.fold_left (+) 0 numbers)

let append : library_function = fun args ->
    let lists = List.map (fun arg -> match arg with
        | List l -> l
        | _ -> raise (RuntimeError "append: invalid arguments")) args in
    List (List.concat lists)

let null : library_function = fun args -> match args with
    | [List []] -> Bool true
    | [List _] -> Bool false
    | _ -> raise (RuntimeError "null?: invalid arguments")

let gt : library_function = fun args -> match args with
    | [Integer lhs; Integer rhs] -> Bool(lhs > rhs)
    | _ -> raise (RuntimeError ">: invalid arguments")

let lt : library_function = fun args -> match args with
    | [Integer lhs; Integer rhs] -> Bool(lhs < rhs)
    | _ -> raise (RuntimeError "<: invalid arguments")

let equals : library_function = fun args -> match args with
    | [Integer lhs; Integer rhs] -> Bool(lhs = rhs)
    | [Bool lhs; Bool rhs] -> Bool(lhs = rhs)
    | [String lhs; String rhs] -> Bool(lhs = rhs)
    | [List lhs; List rhs] ->
        (try
            Bool (List.for_all2 (=) lhs rhs)
        with Invalid_argument _ -> Bool false)
    | [_; _] -> Bool false
    | _ -> raise (RuntimeError "=: invalid arguments")

let functions = [
    "lambda", LibraryMacro lambda;
    "define", LibraryMacro define;
    "set!", LibraryMacro set;
    "let", LibraryMacro letm;
    "if", LibraryMacro ifm;
    "eval", LibraryMacro eval;
    "defmacro", LibraryMacro defmacro;
    "meval", LibraryMacro meval;
    "print", LibraryFunction print;
    "car", LibraryFunction car;
    "cdr", LibraryFunction cdr;
    "cons", LibraryFunction cons;
    "+", LibraryFunction plus;
    "append", LibraryFunction append;
    "null?", LibraryFunction null;
    "<", LibraryFunction lt;
    ">", LibraryFunction gt;
    "=", LibraryFunction equals;
]

let library =
    let env = Environment.new_env () in
    List.iter (fun (name, value) -> Environment.set env name value) functions;
    env

File: scheme/haskell/Library.hs
module Library(library) where

import Ast
import Eval

library :: IO Environment
library = do
    env <- makeEnv Nothing Nothing
    mapM_ (uncurry $ setEnv env) functions
    return env

functions :: [(String, Expr)]
functions = [
    ("lambda", LibraryMacro lambda),
    ("define", LibraryMacro define),
    ("set!", LibraryMacro set),
    ("let", LibraryMacro letm),
    ("if", LibraryMacro ifm),
    ("eval", LibraryMacro evalm),
    ("defmacro", LibraryMacro defmacro),
    ("meval", LibraryMacro meval),
    ("print", LibraryFunction printFn),
    ("car", LibraryFunction car),
    ("cdr", LibraryFunction cdr),
    ("cons", LibraryFunction cons),
    ("+", LibraryFunction plus),
    ("append", LibraryFunction append),
    ("null?", LibraryFunction nullFn),
    ("<", LibraryFunction lt),
    (">", LibraryFunction gt),
    ("=", LibraryFunction equals)]

lambda :: LibraryMacro
lambda (List params:tl) env = return $ Closure params env (StatementList tl)
lambda _ _ = fail "lambda: invalid arguments"

define :: LibraryMacro
define (Var name:tl) env = do
    current <- envHas env name
    if current
        then fail "define: cannot redefine variable"
        else do
            value <- eval (StatementList tl) env
            setEnv env name value
            return value
define _ _ = fail "define: invalid arguments"

set :: LibraryMacro
set (Var name:tl) env = do
    value <- eval (StatementList tl) env
    setEnv env name value
    return value
set _ _ = fail "set!: invalid arguments"

letm :: LibraryMacro
letm (List definitions:expr) env = do
    newEnv <- makeEnv (Just env) Nothing
    mapM_ (defineOne newEnv) definitions
    eval (StatementList expr) newEnv
        where
            defineOne newEnv (List (Var name:tl)) = do
                value <- eval (StatementList tl) env
                setEnv newEnv name value
            defineOne _ _ = fail "let: invalid arguments"
letm _ _ = fail "let: invalid arguments"

ifm :: LibraryMacro
ifm [conditionExpr, ifTrue, ifFalse] env = do
    condition <- eval conditionExpr env
    case condition of
        Bool False -> eval ifFalse env
        _ -> eval ifTrue env
ifm _ _ = fail "if: invalid arguments"

evalm :: LibraryMacro
evalm [expr] env = do
    toEval <- eval expr env
    eval toEval env
evalm _ _ = fail "eval: invalid arguments"

defmacro :: LibraryMacro
defmacro (Var name:List params:tl) definitionEnv = do
    setEnv definitionEnv name (LibraryMacro newMacro)
    return $ LibraryMacro newMacro
        where
            newMacro :: LibraryMacro
            newMacro args callEnv = do
                newEnv <- makeEnv (Just definitionEnv) (Just callEnv)
                setArgs params args newEnv
                eval (StatementList tl) newEnv

meval :: LibraryMacro
meval [expr] env = do
    case mevalEnv env of
        Nothing -> fail "meval: invalid arguments"
        Just env' -> do
            toEval <- eval expr env
            eval toEval env'
meval _ _ = fail "meval: invalid arguments"

printFn :: LibraryFunction
printFn args = do
    mapM_ (\arg -> putStrLn $ show arg) args
    return $ List []

car :: LibraryFunction
car [List(hd:_)] = return hd
car _ = fail "car: invalid arguments"

cdr :: LibraryFunction
cdr [List(_:tl)] = return $ List tl
cdr _ = fail "cdr: invalid arguments"

cons :: LibraryFunction
cons [hd, List tl] = return $ List(hd:tl)
cons _ = fail "cons: invalid arguments"

plus :: LibraryFunction
plus args = do
    result <- plus' args
    return $ Integer result
        where
            plus' (Integer hd:tl) = do
                rest <- plus' tl
                return $ hd + rest
            plus' [] = return 0
            plus' _ = fail "+: invalid arguments"

append :: LibraryFunction
append args = do
    result <- append' args
    return $ List $ concat result
        where
            append' (List hd:tl) = do
                rest <- append' tl
                return $ hd:rest
            append' [] = return []
            append' _ = fail "append: invalid arguments"

nullFn :: LibraryFunction
nullFn [List []] = return $ Bool True
nullFn [List _] = return $ Bool False
nullFn _ = fail "null?: invalid arguments"

gt :: LibraryFunction
gt = integerFn ">" (>)

lt :: LibraryFunction
lt = integerFn "<" (<)

integerFn :: String -> (Int -> Int -> Bool) -> LibraryFunction
integerFn name fn [Integer lhs, Integer rhs] = return $ Bool $ fn lhs rhs
integerFn name _ _ = fail $ name ++ ": invalid arguments"

equals :: LibraryFunction
equals [lhs, rhs] = return $ Bool $ lhs == rhs
equals _ = fail "=: invalid arguments"
